src/coreComponents/physicsSolvers/multiphysics/CoupledReservoirAndWellsBase.cpp

```diff
@@ -0,0 +1,291 @@
+getWellSolver() const { return std::get< toUnderlying< SolverType >( SolverType::Well ) >( m_solvers ); }
+
+template< typename RESERVOIR_SOLVER, typename WELL_SOLVER >
+void CoupledReservoirAndWellsBase< RESERVOIR_SOLVER, WELL_SOLVER >::
+initializePostInitialConditionsPreSubGroups()
+{
+  AbstractBase::initializePostInitialConditionsPreSubGroups( );
+
+  DomainPartition & domain = this->template getGroupByPath< DomainPartition >( "/Problem/domain" );
+
+  this->template forMeshTargets( domain.getMeshBodies(), [&] ( string const &,
+                                                               MeshLevel & meshLevel,
+                                                               arrayView1d< string const > const & regionNames )
+  {
+    ElementRegionManager & elemManager = meshLevel.getElemManager();
+
+    // loop over the wells
+    elemManager.forElementSubRegions< WellElementSubRegion >( regionNames, [&]( localIndex const,
+                                                                                WellElementSubRegion & subRegion )
+    {
+      array1d< array1d< arrayView3d< real64 const > > > const permeability =
+        elemManager.constructMaterialExtrinsicAccessor< PermeabilityBase, extrinsicMeshData::permeability::permeability >();
+
+      PerforationData * const perforationData = subRegion.getPerforationData();
+
+      // compute the Peaceman index (if not read from XML)
+      perforationData->computeWellTransmissibility( meshLevel, subRegion, permeability );
+    } );
+  } );
+}
+
+template< typename RESERVOIR_SOLVER, typename WELL_SOLVER >
+void CoupledReservoirAndWellsBase< RESERVOIR_SOLVER, WELL_SOLVER >::
+addCouplingNumNonzeros( DomainPartition & domain,
+                        DofManager & dofManager,
+                        arrayView1d< localIndex > const & rowLengths ) const
+{
+  integer const resNDOF = getWellSolver()->numDofPerResElement();
+  integer const wellNDOF = getWellSolver()->numDofPerWellElement();
+
+  this->template forMeshTargets( domain.getMeshBodies(), [&] ( string const &,
+                                                               MeshLevel const & meshLevel,
+                                                               arrayView1d< string const > const & regionNames )
+  {
+    ElementRegionManager const & elemManager = meshLevel.getElemManager();
+
+    string const wellDofKey = dofManager.getKey( getWellSolver()->wellElementDofName() );
+    string const resDofKey = dofManager.getKey( getWellSolver()->resElementDofName() );
+
+    ElementRegionManager::ElementViewAccessor< arrayView1d< globalIndex const > > const & resElemDofNumber =
+      elemManager.constructArrayViewAccessor< globalIndex, 1 >( resDofKey );
+
+    ElementRegionManager::ElementViewAccessor< arrayView1d< integer const > > const & resElemGhostRank =
+      elemManager.constructArrayViewAccessor< integer, 1 >( ObjectManagerBase::viewKeyStruct::ghostRankString() );
```

Comment:
Could not you use some `Accessor`?

Reply:
This brings us back to yesterday's discussion on Mark's PR about using accessors for the dofNumbers:
https://github.com/GEOSX/GEOSX/pull/1733#discussion_r860336851
So yes, maybe we should implement the accessors for dynamic keys (if possible?)